assignments/burmanm/week-4.scm

```diff
@@ -0,0 +1,208 @@
+;; ((z (lambda (p q) p)) ((lambda (m) (m x y) x y)))
+;; ((lambda (m) (m x y) x y) (lambda (p q) p))
+;; m is substituted with the (lambda (p q) p) which returns first argument of (x y)
+
+;; Exercise 2.5
+
+;; Math question again.. meh. 
+
+;; Exercise 2.6
+
+;; (define zero (lambda (f) (lambda (x) x)))
+;; (define (add-1 n)
+;;   (lambda (f) (lambda (x) (f ((n f) x)))))
+
+
+;; (add-1 zero)
+
+(define one (lambda (f) (lambda (x) (f ((lambda (x) x) x)))))
+
+;; call the last one, outer lambda with x returns x..
+(define one (lambda (f) (lambda (x) (f x))))
+
+;; (add-1 one) same idea
+(define two (lambda (f) (lambda (x) (f (f x)))))
+
+;; +
+
+;; Is basically calling add-1 n-times. That is, (f ((n f) x)) * n-times. Since n is used, use m:
+
+(define + (x y) (lambda (f) (lambda (x) ((m f) ((n f) x)))))
+
+;; Exercise 2.7
+
+(define (upper-bound interval) (car interval))
+(define (lower-bound interval) (cdr interval))
+
+;; Exercise 2.8
+
+(define (sub-interval x y)
+  (make-interval (- (lower-bound x) (lower-bound y))
+		 (- (upper-bound x) (upper-bound y))))
+
+;; Exercise 2.9
+
+;; More math..
+
+;; w = (x - y)/2
+;; [x,y] ; [z,v]
+;; wC = ((x + z) - (y + v))/2
+;; w2 = (z - v)/2
+
+;; Prove that those are equal.. that is:
+
+;; w + w2 = wC
+
+;; Easier to read from paper, but it transforms to:
+
+;; ((z-v) + (x-y)) / 2 = ((x+z) - (y+v))/2
+;; * 2 on each side, open parentheses:
+
+;; z-v+x-y = x+z-y-v
+;; Reorder and see that 0 = 0
+
+;; Exercise 2.10
+
+(define (div-interval x y)
+  (if (and
+       (<= 0 (lower-bound y))
+       (>= 0 (upper-bound y)))
+      (error "Intervals spans zero")
+
+      (mul-interval
+       x
+       (make-interval (/ 1.0 (upper-bound y))
+		      (/ 1.0 (lower-bound y))))))
+
+;; Exercise 2.11
+
+;; Insanely complex code, which is bad.
+
+;; Exercise 2.12
+
+;; center and width from SICP
+
+(define (make-center-percent center tolerance)
+  (make-center-width center
+		     (* center (/ 100.0 tolerance))))
+
+(define (percent interval)
+  (* 100.0 (/ (width interval) (center interval))))
+
+;; Exercise 2.13
+
+;; The floating point approximations are the source of issues, as 1/(1/R1 + 1/R2) does require significantly more digits to retain the same amount of accuracy as R1R2/(R1+R2).
+
+;; Exercise 2.14
+
+;; scheme@(guile-user) [5]> (par1 (make-center-percent 1000 2) (make-center-percent 2000 1))
+;; $4 = (-41708.5020242915 . 40882.59109311741)
+;; scheme@(guile-user) [5]> (par2 (make-center-percent 1000 2) (make-center-percent 2000 1))
+;; $5 = (-39279.35222672065 . 40719.3675889328)
+;; scheme@(guile-user) [5]>
+
+;; Exercise 2.15
+
+;; To understand why, one has to look at the implementation of div-interval. Since dividing by itself does not return itself, we pretty quickly end up in a situation where the returns are not exact. And with such inexactness, the calculations start to differ very easily - even if mathematically they should be identical.
+
+;; Exercise 2.16
+
+;; Beyond simple exercise.
```

Comment:
About 2.14/15/16, it's kind of disappointing that after all of that, the author shows that the interval package is not that great.